paddle/gserver/dataproviders/PyDataProvider2.cpp

```diff
@@ -377,9 +377,15 @@ class PyDataProvider2 : public DataProvider {
             std::swap(callingContexts_[cid], callingContexts_[0]);
             cid = 0;
           }
+
+          PyObjectPtr front;
+          {
+            std::unique_lock<std::mutex> l(mtx_);
+            front = pop_get_front(callingContexts_);
+          }
           {
             PyGuard g;
-            callingContexts_.pop_front();
```

Comment:
To add item or to remove item to callingContexts_ is always under PyGuard. And getting size of this container always in same thread.  It seems that there is no dead lock here.